index.js

```diff
@@ -1,302 +1,293 @@
-              capabilities: richResponse.capabilities
-            }
-          )));
-        }
-      });
-      if (isRunningInDialogFlow) {
-        return Promise.all(promises);
-      } else {
-        return Promise.all(promises).then(() => convo);
-      }
-
-    }
-  }
-  static prepComplete(obj, action = "close", options) {
-    if (this.isPromise(obj)) {
-      return obj.then(obj => this.prepComplete(obj, action, options));
-    } else {
-      return this.complete(obj, action, options);
-    }
-  }
-  static ask(obj, options) {
-    return this.prepComplete(obj, "ask", options);
-  }
-  static close(obj, options) {
-    return this.prepComplete(obj, "close", options);
-  }
-
-  constructor(obj) {
-    if (!obj) {
-      this.conv = Convo.mockConv();
-    } else {
-      this.conv = obj.conv ? obj.conv : obj;
-    }
-
-    this.clear();
-  }
-  clear() {
-    this._write = [];
-    this._speak = [];
-    this._rich = [];
-    return this;
-  }
-  write(message) {
-    this._write.push(message.toString());
-    return this;
-  }
-  speak(message, writeAlso = true) {
-    this._speak.push(message.toString());
-    if (writeAlso) {
-      this.write(message);
-    }
-    return this;
-  }
-  present(media, capabilities = null, send = null) {
-    this._rich.push({capabilities, send, media});
-    return this;
-  }
-  promise(func) {
-    let r = func(this);
-    if (!Convo.isPromise(r)) {
-      return Promise.resolve(r);
-    }
-    return r;
-  }
-  setAccessToken(token) {
-    if (!conv.user) {
-      conv.user = {};
-    }
-    if (!conv.user.access) {
-      conv.user.access = {};
-    }
-    conv.user.access.token = token;
-    return this;
-  }
-  getContext(context) {
-    if (this.conv && this.conv.contexts) {
-      var rContext;
-      if (this.conv.contexts.get) {
-        rContext = this.conv.contexts.get(context);
-      } else {
-        rContext = this.conv.contexts[context];
-      }
-      if (rContext && rContext.parameters) {
-        return rContext.parameters;
-      }
-    }
-    return {};
-  }
-  setContext(context, lifespan, parameters) {
-    if (this.conv && this.conv.contexts) {
-      if (this.conv.contexts.set) {
-        this.conv.contexts.set(context, lifespan, parameters);
-      } else {
-        this.conv.contexts[context] = {lifespan, parameters};
-      }
-    }
-    return this;
-  }
-  getStorage() {
-    if (this.conv && this.conv.user && this.conv.user.storage) {
-      return this.conv.user.storage;
-    }
-    return {};
-  }
+	static mockConv() {
+		return {
+			surface: {
+				capabilities: {
+					has: () => true
+				}
+			},
+			available: {
+				surfaces: {
+					capabilities: {
+						has: () => true
+					}
+				}
+			},
+			user: {
+				access: {},
+				storage: {}
+			},
+			contexts: {}
+		};
+	}
+
+	static setClassMappings(mappings) {
+		this._clsMappings = mappings;
+	}
+
+	/**
+	 * @private
+	 */
+	static processMethod(type, data) {
```

Comment:
New helper method I created to abstract commonality of all of the static methods below